test/jdk/java/math/BigInteger/MutableBigIntegerShiftTests.java

```diff
@@ -0,0 +1,163 @@
+        }
+    }
+
+    @Test
+    public static void testShift() {
+        shift(ORDER_SMALL);
+        shift(ORDER_MEDIUM);
+    }
+
+    /*
+     * Get a random or boundary-case number. This is designed to provide
+     * a lot of numbers that will find failure points, such as max sized
+     * numbers, empty MutableBigIntegers, etc.
+     *
+     * If order is less than 2, order is changed to 2.
+     */
+    private static MutableBigIntegerBox fetchNumber(int order) {
+        int numType = random.nextInt(8);
+        MutableBigIntegerBox result = null;
+        if (order < 2) order = 2;
+
+        int[] val;
+        switch (numType) {
+            case 0: // Empty
+                result = MutableBigIntegerBox.ZERO;
+                break;
+
+            case 1: // One
+                result = MutableBigIntegerBox.ONE;
+                break;
+
+            case 2: // All bits set in number
+                int numInts = (order + 31) >> 5;
+                int[] fullBits = new int[numInts];
+                for(int i = 0; i < numInts; i++)
+                    fullBits[i] = -1;
+
+                fullBits[0] &= -1 >>> -order;
+                result = new MutableBigIntegerBox(fullBits);
+                break;
+
+            case 3: // One bit in number
+                result = MutableBigIntegerBox.ONE.shiftLeft(random.nextInt(order));
+                break;
+
+            case 4: // Random bit density
+                val = new int[(order + 31) >> 5];
+                int iterations = random.nextInt(order);
+                for (int i = 0; i < iterations; i++) {
+                    int bitIdx = random.nextInt(order);
+                    val[bitIdx >> 5] |= 1 << bitIdx;
+                }
+                result = new MutableBigIntegerBox(val);
+                break;
+            case 5: // Runs of consecutive ones and zeros
+                result = ZERO;
+                int remaining = order;
+                int bit = random.nextInt(2);
+                while (remaining > 0) {
+                    int runLength = Math.min(remaining, random.nextInt(order));
+                    result = result.shiftLeft(runLength);
+                    if (bit > 0)
+                        result = result.add(ONE.shiftLeft(runLength).subtract(ONE));
+                    remaining -= runLength;
+                    bit = 1 - bit;
+                }
+                break;
+            case 6: // random bits with trailing space
+                int len = random.nextInt((order + 31) >> 5) + 1;
+                int offset = random.nextInt(len);
+                val = new int[len << 1];
+                for (int i = 0; i < val.length; i++)
+                    val[i] = random.nextInt();
+                result = new MutableBigIntegerBox(val, offset, len);
+                break;
+            default: // random bits
+                result = new MutableBigIntegerBox(new BigInteger(order, random));
+        }
+
+        return result;
+    }
```

Comment:
I think this is getting too complex for a test, there's a risk that it is incorrect...

Is there perhaps a way to have simpler code?

Reply:
The test cases of the code are the same of the test class for `BigInteger.shiftLeft`.

Reply:
I see (with obvious adaptations and one more `case`).

Reply:
The implementation of the shift methods in `MutableBigInteger` has several control flow paths, depending on whether a new `int[]` is needed, whether `arraycopy()` can be used, etc.

Relying on random tests is good if the "search space" (the number of control flow paths) is too big for a systematic approach. But here I think it should be possible to exercise all paths with a dozen or so well targeted tests, in addition to the existing ones.

These are white-box tests, so the internal structure of the algorithms is supposed to be known and this knowledge should be exploited in trying to hit the weak points, if there are.

Reply:
@fabioromano1 So, a test should be done for every possible path of computation of `MBI.leftShift()` method...

Reply:
I think that specific test cases for every path would make the code more reliable.